Allow the store user to delete coupons.

cdepillabout opened the pull request:
This PR allows the `Store` user to delete `Coupon`s.

If you go to a store user coupon page, there is a new delete button at the bottom.

http://localhost:8101/store/coupon/13

Clicking that takes you to the coupon delete page:

http://localhost:8101/store/coupon/delete/13

@arowM There are three things I'd appreciate if you'd take a look at.

1.  I added a 404 page.
    
    https://github.com/blueimpact/kucipong/pull/155/files#diff-31376d72c1f6a4671661d54e0773c618
    https://github.com/blueimpact/kucipong/pull/155/files#diff-66a030d00a32f20aa50ae4fbae9ed27eR384
    
    I think this will simplify some of the handlers for truely exceptional conditions.  For instance, the `Store` user will probably never go to the coupon delete page for a coupon that has already been deleted.  I think it makes sense to just return a 404 page here.

    There are also a couple other handlers in the app that could probably just return a 404 page (instead of showing the user an error message).

2.  The coupon delete page looks a little funny.  If you have any good ideas for making it look better, I'd appreciate it.
    
    https://github.com/blueimpact/kucipong/pull/155/files#diff-14cfbdf62dc35a70414bf93a359c8f98

3.  Since the coupon id template (`frontend/src/pug/common/_coupon_id.pug`) is the same for the end user and store user, I had to add an ugly hack to determine whether or not to show the coupon delete button.  Can you think of any better way of doing this?
    
    https://github.com/blueimpact/kucipong/pull/155/files#diff-38506ea977d44a1bbae213029893f447R85
    https://github.com/blueimpact/kucipong/pull/155/files#diff-d9ff201ce1e331481f5cf6639f82d4dfR8

cdepillabout commented:
This fixes #133.

arowM commented on src/Kucipong/Handler/Store/Coupon.hs:
```diff
@@ -366,6 +372,31 @@ couponPost = do
           , "defaultImage"
           ])
 
+-- | Return the store create page for an admin.
```
Though it's trivial, this seems to be "Return the store delete page for an admin.".

cdepillabout replied:
This has been fixed in commit e0d576f.

arowM commented:
> 2. The coupon delete page looks a little funny. If you have any good ideas for making it look better, I'd appreciate it.

In the view point of [OOUX](https://www.sociomedia.co.jp/7279), it is better to provide a `X` button on each coupon of coupon list page (`/store/coupon`) instead of providing an independent coupon delete page. I'll do this change in future PR.
It should use Ajax, so I'm appreciate if you provide a JSON web API to delete a coupon with its coupon ID.

> 3. Since the coupon id template (frontend/src/pug/common/_coupon_id.pug) is the same for the end user and store user, I had to add an ugly hack to determine whether or not to show the coupon delete button. Can you think of any better way of doing this?

The way you did is not so bad, but I have better ways.
The important point is we need not ask backend server to determine whether we should render delete button or not, but we can determine it in compile time.
So, I have two plan to realize this.
1. Use pug-side variable to handle appearance of delete button
2. Set the style of store delete button `display: block` in `store.scss`.
Anyway since I'll implement OOUX way of delete feature in future PR, then I'll change this, too.

cdepillabout commented:
Thanks for these two good ideas.  I've added issue #156 for me to implement the JSON API for deleting a coupon.  I've also added #157 for @arowM for adding the `X` button on each coupon.
